# Project 5 -- [Niharika] [Rai]

**TA Help:** Sharan Sivakumar. (Question 4a). 
    
**Collaboration:** N/A 

## Question 1

In [1]:
def avg_aggreg_temp(file_location, column_title_list, start_date, end_date, element_code = "TAVG"):
    import pandas as pd 
    myDF = pd.read_csv(file_location, header=None, names=column_title_list) 
    myresult = myDF[(myDF['date'] >= start_date) & (myDF['date'] <= end_date) & (myDF['element_code'] == element_code)]['value'].mean() 
    return myresult 

In [2]:
avg_aggreg_temp("/anvil/projects/tdm/data/noaa/1880.csv", ["id","date","element_code","value","mflag","qflag","sflag","obstime"], 18801001, 18801031, "TAVG") 

161.06451612903226

In [1]:
def avg_aggreg_temp(file_location, column_title_list, start_date, end_date, element_code = "TAVG"):
    import pandas as pd
    myDF = pd.read_csv(file_location, header=None, names=column_title_list) 
    myresult = myDF[(myDF['date'] >= start_date) & (myDF['date'] <= end_date) & (myDF['element_code'] == element_code)]['value'].mean() 
    return myresult 

In [2]:
avg_aggreg_temp("/anvil/projects/tdm/data/noaa/2018.csv", ["id","date","element_code","value","mflag","qflag","sflag","obstime"], 20180101, 20180115, "TAVG") 

12.317542410053962

For question 1, part a, I used the video, "how to write a function for analyzing the NOAA data in a given date range" given by Dr. Ward: https://the-examples-book.com/programming-languages/python/writing-functions-page-2 to get the code structure for writing a function and checking it with the 1880.csv file to ensure that I also got the same answer, which meant that my fuction works. For part b, I used the same code structure but replaced the year, start_date and end_date with the values given in the question. I did not use chunksize for this question. Note: I read the Piazza discussion about the minimum number of cores being 4 and my kernel still died. So, I set my cores to 6 to get the output for the 2018.csv file. 

## Question 2

In [6]:
def avg_aggreg_temp_by_year(range_of_years, column_title_list, element_code = "TAVG"):
    import pandas as pd 
    
    mydict = {} 
    for year in range_of_years:
        file_location = f'/anvil/projects/tdm/data/noaa/{year}.csv' 
        myDF = pd.read_csv(file_location, header=None, names=column_title_list) 
        myresult = myDF[(myDF['element_code'] == element_code)]['value'].mean()
        mydict[year] = myresult
    return mydict 

avg_aggreg_temp_by_year(range(1880, 1884), ["id","date","element_code","value","mflag","qflag","sflag","obstime"], element_code = "PRCP") 

{1880: 17.528739659455965,
 1881: 16.05386765236206,
 1882: 16.258667810567434,
 1883: 15.521981600238885}

In [7]:
avg_aggreg_temp_by_year(range(1880, 1884), ["id","date","element_code","value","mflag","qflag","sflag","obstime"], element_code = "TAVG") 

{1880: 127.30054644808743,
 1881: 49.00669116954104,
 1882: 58.33541290428501,
 1883: 50.54050116550116}

For parts a and b, I watched the video, "returning a dictionary with the average values for each year" in this link: https://the-examples-book.com/programming-languages/python/writing-functions-page-2. Hence, the video provided me with the code structure which I later modified for "PRCP" to get a specified element_code value. For part b, I replaced the "PRCP" with "TAVG" to get the appropriate dictionary values. 

## Question 3

In [80]:
def avg_aggreg_temp_by_year_and_month(range_of_years, mymonth, column_title_list, element_code = "TAVG"):
    import pandas as pd 
    
    mydict = {} 
    for year in range_of_years:
        file_location = f'/anvil/projects/tdm/data/noaa/{year}.csv'
        myDF = pd.read_csv(file_location, header=None, names=column_title_list)
        myDF['month'] = pd.to_datetime(myDF['date'], format='%Y%m%d').dt.month
        myresult = myDF[(myDF['element_code'] == element_code) & (myDF['month'] == mymonth)]['value'].mean()
        mydict[year] = myresult
    return mydict 

avg_aggreg_temp_by_year_and_month(range(1880, 1884), 8, ["id","date","element_code","value","mflag","qflag","sflag","obstime"], element_code = "TAVG")

{1880: 225.70967741935485,
 1881: 188.42227122381476,
 1882: 207.34585741811176,
 1883: 185.22810060711188}

For part a and b, I got the code structure from the previous question. However, I added the month parameter and an additional code line that extracts the data by month by watching the video, "taking an average for the data in the selected month" in this link: https://the-examples-book.com/programming-languages/python/writing-functions-page-2. 

## Question 4

In [21]:
for year in range(1880, 1884):
    import pandas as pd
    from pathlib import Path 
    myDF = pd.read_csv(Path(f'/anvil/projects/tdm/data/noaa/{year}.csv'), header=None, names=["id","date","element_code","value","mflag","qflag","sflag","obstime"])
    mycount = myDF[myDF['qflag'] == 'D'].shape[0] 
    print(f'{year}:{mycount}') 

1880:548
1881:454
1882:365
1883:214


In [ ]:
max_count = 0
max_year = None
for year in range(1880, 1884):
    import pandas as pd
    from pathlib import Path 
    myDF = pd.read_csv(Path(f'/anvil/projects/tdm/data/noaa/{year}.csv'), header=None, names=["id","date","element_code","value","mflag","qflag","sflag","obstime"])  
    mycount = myDF[myDF['qflag'] == 'D'].shape[0]
    if mycount > max_count:
        max_count = mycount
        max_year = year
print(max_year) 

1880


In [34]:
def year_with_most_qflags(list_of_years, qflag):
    max_count = 0
    max_year = None
    for year in list_of_years:
        import pandas as pd
        from pathlib import Path
        myDF = pd.read_csv(Path(f'/anvil/projects/tdm/data/noaa/{year}.csv'), header=None, names=["id","date","element_code","value","mflag","qflag","sflag","obstime"])  
        mycount = myDF[myDF['qflag'] == qflag].shape[0]
        if mycount > max_count:
            max_count = mycount
            max_year = year
    return(max_year) 

year_with_most_qflags([1880, 1881, 1882, 1883], 'D') 

1880

In [35]:
year_with_most_qflags([1880, 1881, 1882, 1883], 'G')

1880

In [36]:
year_with_most_qflags([1880, 1881, 1882, 1883], 'I')

1883

In [37]:
year_with_most_qflags([1880, 1881, 1882, 1883], 'K')

1882

In [38]:
year_with_most_qflags([1880, 1881, 1882, 1883], 'L')

1881

In [39]:
year_with_most_qflags([1880, 1881, 1882, 1883], 'N')

1880

In [40]:
year_with_most_qflags([1880, 1881, 1882, 1883], 'O')

1880

In [41]:
year_with_most_qflags([1880, 1881, 1882, 1883], 'S')

1882

In [42]:
year_with_most_qflags([1880, 1881, 1882, 1883], 'X')

1881

For part a, I first got the mycount code structure from Dr.Ward through Piazza. From the output of this code, I knew that 548 is the maximum count for "D". So, I needed to create a code that gave 1880 as the output. This confused me. So, I attended office hours and got a TA's help to figure out how to get the code structure. His name is Sharan Sivakumar. He helped me understand the meaning of the if statement that I made to identify the max_year and max_count. He also helped me assign the values to max_count and max_year initally and helped me understand what those meant. When my second code gave me the output as 1880, which was what I wanted, I knew that I can now create the function. So, then I defined my fuction based on the parameters asked by the question. However, I had to make a minor change. Instead of, " mycount = myDF[myDF['qflag'] == 'D'].shape[0]", I switched the "D" value with "qflag". So, my code now look like this: mycount = myDF[myDF['qflag'] == qflag].shape[0]. Then, I inputed the list of years and the specific qflag values, for part b, to test out the function. And, it gave me the intended outputs. 

## Question 5

In [6]:
import pandas as pd
myDF = pd.read_csv("/anvil/projects/tdm/data/noaa/1800.csv", header=None, names=["id","date","element_code","value","mflag","qflag","sflag","obstime"]) 
myDF.head() 

,id,date,element_code,value,mflag,qflag,sflag,obstime
0,EZE00100082,18000101,TMAX,-86,NaN,NaN,E,NaN
1,EZE00100082,18000101,TMIN,-135,NaN,NaN,E,NaN
2,GM000010962,18000101,PRCP,0,NaN,NaN,E,NaN
3,ITE00100554,18000101,TMAX,-75,NaN,NaN,E,NaN
4,ITE00100554,18000101,TMIN,-148,NaN,NaN,E,NaN


In [44]:
def avg_min_temp(file_location, column_title_list, element_code = "TAVG"):
    
    """
    avg_min_temp is a function that accepts the file_location, column_title_list and element_code = "TAVG" as arguements and returns the mean of the specified element_code value. 
    
    Args: 
    
    file_location: This parameter is the file location from which the data will be extracted. In this question, I used the 1800.csv file. 
    column_title_list: This parameter helps to name the columns in the 1800.csv file because all the files in the noaa directory are unnamed. 
    element_code = "TAVG": This parameter is one of the column names where "TAVG" is the default value. When the element_code is specfied, the fuction will use that specified value to return the output. 
    
    Returns: 
    
    myresult: This variable has been designed to extract the mean when the element_code is assigned a specific row such as "PRCP", "TMAX", "TMIN" etc. If no row name has been specified, then "TAVG" will be taken as the default row. 
    
    Since I set the element_code value to "TMIN", the function will read through the data file and pull out all the numerical values where the row has "TMIN" and take the average of those values. 
    
    So, the output of my function communicates that average_min_temp (average minimum temperature) to be 90.84794520547945 in the 1800.csv file. 
    
    
    Note: the 2 statements above myresult are to ensure that the data file is read properly. Otherwise the function won't be able to extract the values needed to give the output. 
    """
    import pandas as pd 
    myDF = pd.read_csv(file_location, header=None, names=column_title_list) 
    myresult = myDF[(myDF['element_code'] == element_code)]['value'].mean() 
    return myresult 

avg_min_temp("/anvil/projects/tdm/data/noaa/1800.csv", ["id","date","element_code","value","mflag","qflag","sflag","obstime"], "TMIN") 

90.84794520547945

In [45]:
help(avg_min_temp) 

Help on function avg_min_temp in module __main__:

avg_min_temp(file_location, column_title_list, element_code='TAVG')
    avg_min_temp is a function that accepts the file_location, column_title_list and element_code = "TAVG" as arguements and returns the mean of the specified element_code value. 
    
    Args: 
    
    file_location: This parameter is the file location from which the data will be extracted. In this question, I used the 1800.csv file. 
    column_title_list: This parameter helps to name the columns in the 1800.csv file because all the files in the noaa directory are unnamed. 
    element_code = "TAVG": This parameter is one of the column names where "TAVG" is the default value. When the element_code is specfied, the fuction will use that specified value to return the output. 
    
    Returns: 
    
    myresult: This variable has been designed to extract the mean when the element_code is assigned a specific row such as "PRCP", "TMAX", "TMIN" etc. If no row name has 

For question 5, I selected the 1800.csv file and I got the code structure from question 1. However, I modified the code to explore some newer values. So, instead of "PRCP", I used "TMIN". But before that, I read the 1800.csv file and used the .head() funtion to find out what other row names exist under the element_code column name. Most importantly, I used the 1800.csv file because it worked with 1 core. When I tried to read other files, my kernel would always die. So, I found the 1800.csv file to be an easier option where I could explore things that would give proper results without crashing. I read about docstrings from this page in the examples book: https://the-examples-book.com/programming-languages/python/writing-functions#docstrings. I then, imitated the same structure. So, I inserted the 3 quotes (""") below the fuction and 3 quotes (""") just above the import pandas as pd statement. In between these quotes was where I wrote my description of the fuction, each of its parameters, how the output is returned and what it means in context of the 1800.csv file. To check if the docstring works, I used the help() fuction to extract out the docstring. I got this code structure also from the same textbook page, which referenced the example: help(word_count). So, I replaced "word_count" with my function's name, "avg_min_temp" to get the documentation I wrote to explain how my function worked. 

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.